kv/kvserver: TestLeaseholdersRejectClockUpdateWithJump failed

[(kv/kvserver).TestLeaseholdersRejectClockUpdateWithJump failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2702602&tab=buildLog) on [master@bf9744bad5a416a4b06907f0f3dd42896f7342f3](https://github.com/cockroachdb/cockroach/commits/bf9744bad5a416a4b06907f0f3dd42896f7342f3):

```
=== RUN   TestLeaseholdersRejectClockUpdateWithJump
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump291996701
    test_log_scope.go:74: use -show-logs to present logs inline
    client_replica_test.go:266: remote wall time is too far ahead (500.00652ms) to be trustworthy
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump291996701
--- FAIL: TestLeaseholdersRejectClockUpdateWithJump (1.57s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=4

```
make stressrace TESTS=TestLeaseholdersRejectClockUpdateWithJump PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestLeaseholdersRejectClockUpdateWithJump.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestLeaseholdersRejectClockUpdateWithJump failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2709011&tab=buildLog) on [master@43174600d4e72030e9d71792c09935118f6e26ff](https://github.com/cockroachdb/cockroach/commits/43174600d4e72030e9d71792c09935118f6e26ff):

```
=== RUN   TestLeaseholdersRejectClockUpdateWithJump
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump432958884
    test_log_scope.go:74: use -show-logs to present logs inline
    client_replica_test.go:266: remote wall time is too far ahead (500.00107ms) to be trustworthy
E210223 07:23:35.159548 451980 kv/kvserver/consistency_queue.go:191  [n1,consistencyChecker,s1,r10/1:/Table/1{4-5}] 1  computing own checksum: could not dial node ID 1: failed to connect to n1 at 127.0.0.1:40325: stopped
E210223 07:23:35.161010 451980 kv/kvserver/queue.go:1093  [n1,consistencyChecker,s1,r10/1:/Table/1{4-5}] 2  computing own checksum: could not dial node ID 1: failed to connect to n1 at 127.0.0.1:40325: stopped
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump432958884
--- FAIL: TestLeaseholdersRejectClockUpdateWithJump (1.83s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestLeaseholdersRejectClockUpdateWithJump PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestLeaseholdersRejectClockUpdateWithJump.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestLeaseholdersRejectClockUpdateWithJump failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2770128&tab=buildLog) on [master@4b98115dfda02a9498f566958bd915c45ec7e449](https://github.com/cockroachdb/cockroach/commits/4b98115dfda02a9498f566958bd915c45ec7e449):

```
=== RUN   TestLeaseholdersRejectClockUpdateWithJump
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump884757335
    test_log_scope.go:74: use -show-logs to present logs inline
    client_replica_test.go:266: remote wall time is too far ahead (500.000105ms) to be trustworthy
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump884757335
--- FAIL: TestLeaseholdersRejectClockUpdateWithJump (0.16s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestLeaseholdersRejectClockUpdateWithJump PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestLeaseholdersRejectClockUpdateWithJump.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestLeaseholdersRejectClockUpdateWithJump failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2771845&tab=buildLog) on [master@c3ab7bc616e7228b00a00cc25812f7cbbd576404](https://github.com/cockroachdb/cockroach/commits/c3ab7bc616e7228b00a00cc25812f7cbbd576404):

```
=== RUN   TestLeaseholdersRejectClockUpdateWithJump
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump833353221
    test_log_scope.go:74: use -show-logs to present logs inline
    client_replica_test.go:266: remote wall time is too far ahead (503.516348ms) to be trustworthy
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump833353221
--- FAIL: TestLeaseholdersRejectClockUpdateWithJump (0.35s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestLeaseholdersRejectClockUpdateWithJump PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestLeaseholdersRejectClockUpdateWithJump.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestLeaseholdersRejectClockUpdateWithJump failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2789400&tab=buildLog) on [master@36dea46f8cedf42df31b57dd70db7e0f1fd7a453](https://github.com/cockroachdb/cockroach/commits/36dea46f8cedf42df31b57dd70db7e0f1fd7a453):

```
=== RUN   TestLeaseholdersRejectClockUpdateWithJump
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump828508967
    test_log_scope.go:74: use -show-logs to present logs inline
    client_replica_test.go:266: remote wall time is too far ahead (500.00005ms) to be trustworthy
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump828508967
--- FAIL: TestLeaseholdersRejectClockUpdateWithJump (0.31s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestLeaseholdersRejectClockUpdateWithJump PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestLeaseholdersRejectClockUpdateWithJump.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestLeaseholdersRejectClockUpdateWithJump failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2792797&tab=buildLog) on [master@ee9f47b9ec9476a693464e2dcd09a01bf9d39ad2](https://github.com/cockroachdb/cockroach/commits/ee9f47b9ec9476a693464e2dcd09a01bf9d39ad2):

```
=== RUN   TestLeaseholdersRejectClockUpdateWithJump
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump188507139
    test_log_scope.go:74: use -show-logs to present logs inline
    client_replica_test.go:266: remote wall time is too far ahead (500.000104ms) to be trustworthy
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump188507139
--- FAIL: TestLeaseholdersRejectClockUpdateWithJump (0.18s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestLeaseholdersRejectClockUpdateWithJump PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestLeaseholdersRejectClockUpdateWithJump.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestLeaseholdersRejectClockUpdateWithJump failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2855393&tab=buildLog) on [master@651184b3cfbd7c6d336bb49e7152e3d48ea06416](https://github.com/cockroachdb/cockroach/commits/651184b3cfbd7c6d336bb49e7152e3d48ea06416):

```
=== RUN   TestLeaseholdersRejectClockUpdateWithJump
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump259854745
    test_log_scope.go:74: use -show-logs to present logs inline
    client_replica_test.go:266: remote wall time is too far ahead (500.000081ms) to be trustworthy
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump259854745
--- FAIL: TestLeaseholdersRejectClockUpdateWithJump (0.19s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestLeaseholdersRejectClockUpdateWithJump PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestLeaseholdersRejectClockUpdateWithJump.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[kv/kvserver.TestLeaseholdersRejectClockUpdateWithJump failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2869057&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2869057&tab=artifacts#/) on [master @ 71a023f9212cf1285a4475ed85d5ab6eefa9bdf7](https://github.com/cockroachdb/cockroach/commits/71a023f9212cf1285a4475ed85d5ab6eefa9bdf7):


```
=== RUN   TestLeaseholdersRejectClockUpdateWithJump
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump051894853
    test_log_scope.go:74: use -show-logs to present logs inline
    client_replica_test.go:266: remote wall time is too far ahead (500.000428ms) to be trustworthy
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump051894853
--- FAIL: TestLeaseholdersRejectClockUpdateWithJump (0.20s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestLeaseholdersRejectClockUpdateWithJump PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</p>
</details>

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseholdersRejectClockUpdateWithJump.*&sort=title+created&display=lastcommented+project)

<sub>[Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub>

kv/kvserver.TestLeaseholdersRejectClockUpdateWithJump [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2901058&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2901058&tab=artifacts#/) on master @ [d094f020d3ccb52ea5c209b3e124395fb72a03bb](https://github.com/cockroachdb/cockroach/commits/d094f020d3ccb52ea5c209b3e124395fb72a03bb):


```
=== RUN   TestLeaseholdersRejectClockUpdateWithJump
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump340127099
    test_log_scope.go:74: use -show-logs to present logs inline
    client_replica_test.go:266: remote wall time is too far ahead (500.00011ms) to be trustworthy
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump340127099
--- FAIL: TestLeaseholdersRejectClockUpdateWithJump (0.17s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestLeaseholdersRejectClockUpdateWithJump PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</p>
</details>
/cc @nvanbenschoten
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseholdersRejectClockUpdateWithJump.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestLeaseholdersRejectClockUpdateWithJump [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2923864&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2923864&tab=artifacts#/) on master @ [3fcf9313c943c1982133266aa6a08f10c78a19c9](https://github.com/cockroachdb/cockroach/commits/3fcf9313c943c1982133266aa6a08f10c78a19c9):


```
=== RUN   TestLeaseholdersRejectClockUpdateWithJump
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump051231210
    test_log_scope.go:74: use -show-logs to present logs inline
    client_replica_test.go:266: remote wall time is too far ahead (500.00437ms) to be trustworthy
E210426 15:30:03.167855 719300 server/status.go:1264  [n1,status] 1  aborted during DistSender.Send: context canceled
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLeaseholdersRejectClockUpdateWithJump051231210
--- FAIL: TestLeaseholdersRejectClockUpdateWithJump (1.66s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestLeaseholdersRejectClockUpdateWithJump PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @nvanbenschoten
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseholdersRejectClockUpdateWithJump.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/util/hlc/hlc.go
+++ b/pkg/util/hlc/hlc.go
@@ -128,8 +128,7 @@ func (m *ManualClock) Set(nanos int64) {
 // ticks with the wall clock, but that can be moved arbitrarily
 // into the future or paused. HybridManualClock is thread safe.
 type HybridManualClock struct {
-	physicalClock func() int64
-	mu            struct {
+	mu struct {
 		syncutil.RWMutex
 		// nanos, if not 0, is the amount of time the clock was manually incremented
 		// by; it is added to physicalClock.
@@ -143,7 +142,7 @@ type HybridManualClock struct {
 // NewHybridManualClock returns a new instance, initialized with
 // specified timestamp.
 func NewHybridManualClock() *HybridManualClock {
-	return &HybridManualClock{physicalClock: UnixNano}
+	return &HybridManualClock{}
 }
 
 // UnixNano returns the underlying hybrid manual clock's timestamp.
@@ -155,7 +154,7 @@ func (m *HybridManualClock) UnixNano() int64 {
 	if nanosAtPause > 0 {
 		return nanos + nanosAtPause
 	}
-	return nanos + m.physicalClock()
+	return nanos + UnixNano()
 }
 
 // Increment increments the hybrid manual clock's timestamp.
@@ -169,7 +168,7 @@ func (m *HybridManualClock) Increment(nanos int64) {
 // the clock to tick. Increment can still be used, though.
 func (m *HybridManualClock) Pause() {
 	m.mu.Lock()
-	m.mu.nanosAtPause = m.physicalClock()
+	m.mu.nanosAtPause = UnixNano()
 	m.mu.Unlock()
 }
 

--- a/pkg/util/hlc/hlc.go
+++ b/pkg/util/hlc/hlc.go
@@ -128,38 +128,49 @@ func (m *ManualClock) Set(nanos int64) {
 // ticks with the wall clock, but that can be moved arbitrarily
 // into the future or paused. HybridManualClock is thread safe.
 type HybridManualClock struct {
-	nanos         int64
 	physicalClock func() int64
-	// nanosAtPause records the timestamp of the clock if it was paused. A 0 value
-	// indicates the clock was never paused.
-	nanosAtPause int64
+	mu            struct {
+		syncutil.RWMutex
+		// nanos, if not 0, is the amount of time the clock was manually incremented
+		// by; it is added to physicalClock.
+		nanos int64
+		// nanosAtPause records the timestamp of the physical clock when it gets
+		// paused. 0 means that the clock is not paused.
+		nanosAtPause int64
+	}
 }
 
 // NewHybridManualClock returns a new instance, initialized with
 // specified timestamp.
 func NewHybridManualClock() *HybridManualClock {
-	return &HybridManualClock{nanos: 0, physicalClock: UnixNano, nanosAtPause: 0}
+	return &HybridManualClock{physicalClock: UnixNano}
 }
 
 // UnixNano returns the underlying hybrid manual clock's timestamp.
 func (m *HybridManualClock) UnixNano() int64 {
-	nanosAtPause := atomic.LoadInt64(&m.nanosAtPause)
-	nanos := atomic.LoadInt64(&m.nanos)
+	m.mu.RLock()
+	nanosAtPause := m.mu.nanosAtPause
+	nanos := m.mu.nanos
+	m.mu.RUnlock()
 	if nanosAtPause > 0 {
 		return nanos + nanosAtPause
 	}
 	return nanos + m.physicalClock()
 }
 
-// Increment atomically increments the hybrid manual clock's timestamp.
-func (m *HybridManualClock) Increment(incr int64) {
-	atomic.AddInt64(&m.nanos, incr)
+// Increment increments the hybrid manual clock's timestamp.
+func (m *HybridManualClock) Increment(nanos int64) {
+	m.mu.Lock()
+	m.mu.nanos += nanos
+	m.mu.Unlock()
 }
 
-// Pause pauses the hybrid manual clock and forces it to always return the
-// current timestamp.
+// Pause pauses the hybrid manual clock; the passage of time no longer causes
+// the clock to tick. Increment can still be used, though.
 func (m *HybridManualClock) Pause() {
-	atomic.StoreInt64(&m.nanosAtPause, m.physicalClock())
+	m.mu.Lock()
+	m.mu.nanosAtPause = m.physicalClock()
+	m.mu.Unlock()
 }
 
 // UnixNano returns the local machine's physical nanosecond

